### DISCUSSION TITLE
Display error when logging with wrong provider [credentials]

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#14622] [Archive]
Hi,

There was a recent improvement thanks to @aidenfoxx https://github.com/directus/directus/pull/13184/files about displaying an error message when we try to authentify with the wrong provider.

Shouldn't it work the same also when the account has been created with credentials and we try to login with another provider ?

Here is a quite usual flow:
- I create an account with credentials
- I try to login with any other providers

I would expect all the providers would behave the same and so to have also the error `INVALID_PROVIDER`.

Currently in case we have activated the public registration, we would have the duplicate account error.

Thanks for feedback

**Comment by aidenfoxx** [comment]
It should work the way you describe. If you log in with the incorrect provider for an account you should get the following error for all flows:

![image](https://user-images.githubusercontent.com/5548518/180754513-7aad6b8c-f9ad-4375-9f3c-0d2f8aba6867.png)

If it's not working as you expect could you please provide detailed steps on how to reproduce the issue?

**Reply by gkielwasser** [reply]
Hi,

To reproduce :
- Create an account with credentials
- Login with an openid provider
I obtain the following exception code `RECORD_NOT_UNIQUE`

The following works as expected:
- Create an account with an openid provider
- Login with another open provider
I obtain the following exception code `INVALID_PROVIDER`

The `fetchuserId` function relying on `external_identifier` won't find the user https://github.com/directus/directus/blob/16fbe743fe68093f6931ede4aebce487e9b3aa3e/api/src/auth/drivers/openid.ts#L106

**Reply by aidenfoxx** [reply]
I'm not 100% sure how to handle this, because it's not actually the provider that's wrong. Just that there is a user in the system with the same email. The odds are that the authenticating user will have an account registered, but it's not guaranteed (for example, if someone spelt their email wrong).

The `INVALID_PROVIDER` error is based on the `external_identifier` and is guaranteed to be correct.

That being said, we could just assume the user is already registered and display the same "User belongs to a different auth provider" message. But it may be better to have a specific error along the lines of "Could not register user. A user with the same email already exists". 

Thoughts @rijkvanzanten?

**Reply by aidenfoxx** [reply]
And just to clarify, I believe this can only happen when the "external_identifier" differs between two auth provider. Is that correct @gkielwasser? So one provider uses "email" and the other uses "username", for example.

**Reply by gkielwasser** [reply]
@aidenfoxx in the case I described It is as following

The account created with credentials has ```{external_identifier: undefined, provider: 'Default'}```
The account created with openid has ```{external_identifier: email, provider: 'Google'}```

So I would more say one has a defined external_identifier whereas the other not

**Reply by aidenfoxx** [reply]
So the local authentication provider doesn't use `external_identifier`, so it is as I suspected. Let's here what Rijk thinks. Probably best to add a new error message I would think.

**Reply by rijkvanzanten** [reply]
> That being said, we could just assume the user is already registered and display the same "User belongs to a different auth provider" message. But it may be better to have a specific error along the lines of "Could not register user. A user with the same email already exists".
>
> Thoughts @rijkvanzanten?

I think the two can be considered equal. The registration piece of SSO login is hidden to the user, so they're not even aware they're registering or logging in. I therefore don't think there's a need for a slightly different error message to the end user. That being said, we might want to log it differently in the logger, as it is important differentiation for administrators 👍🏻

**Reply by aidenfoxx** [reply]
This should solve things https://github.com/directus/directus/pull/14625
